document/js-api/index.bs

```diff
@@ -1149,27 +1143,42 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
             1. Return [=ref.null=] |heaptype|.
         1. Otherwise,
             1. Throw a {{TypeError}}.
-    1. If |type| is a subtype of [=ref=] |null| [=heap-type/func=],
+    1. If |heaptype| is a subtype of [=heap-type/func=],
         1. If |v| is an [=Exported Function=],
             1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
             1. Let |testresult| be [=ref_test=](|heaptype|, [=ref.func=] |funcaddr|).
             1. If |testresult| is true,
                 1. Return [=ref.func=] |funcaddr|.
         1. Throw a {{TypeError}}.
-    1. If |heaptype| is [=extern=],
-        1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-        1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
-            1. Return [=ref.extern=] |externaddr|.
-        1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
-        1. [=map/Set=] |map|[|externaddr|] to |v|.
-        1. Return [=ref.extern=] |externaddr|.
-    1. If |type| is a subtype of [=ref=] |null| [=any=],
-        1. Let |externref| be [=ToWebAssemblyValue=](|v|, [=ref=] [=extern=]).
-        1. Let |internref| be [=extern_internalize=](|externref|).
-        1. Let |testresult| be [=ref_test=](|heaptype|, |internref|).
-        1. If |testresult| is true,
-            1. Return |internref|.
-        1. Throw a {{TypeError}}.
+    1. If |heaptype| is a subtype of [=any=],
+        1. If |v| [=is a Number=],
+            1. Let |i32| be ? [$ToInt32$](|v|).
+            1. If |v| is equal to |i32| and [=ℝ=](|v|) < 2<sup>30</sup> and [=ℝ=](|v|) ⩾ -2<sup>30</sup>,
+                1. If |heaptype| is not a supertype of [=i31=],
+                    1. Throw a {{TypeError}}.
+                1. Return [=ref.i31=] |v|.
+        1. If |v| is an [=Exported GC Object=],
+            1. Let |objectaddr| be the value of |v|'s \[[ObjectAddress]] internal slot.
+            1. Let |objecttype| be the value of |v|'s \[[ObjectType]] internal slot.
+            1. If |objecttype| is "array",
+                1. If |heaptype| is not a supertype of [=array=],
```

Comment:
Supertype or subtype, right? Since we want to do implicit downcasts on the boundary. Same for structs below.

We also need to handle the possibility of the downcast failing for structs, arrays, and functions.

Reply:
This is the check for the downcast. The heaptype is the target type here, so the actual type must be a subtype of it for that cast to succeed.

Using "matching" terminology now, which may make this somewhat clearer.